mentat/code_context.py

```diff
@@ -271,37 +273,133 @@ def _get_all_features(
-    def include_file(self, path: Path):
-        paths, invalid_paths = get_include_files([path], [])
-        for new_path, new_features in paths.items():
-            if new_path not in self.include_files:
-                self.include_files[new_path] = []
-            for feature in new_features:
-                self.include_files[new_path].append(feature)
-        return list(paths.keys()), invalid_paths
-
-    def exclude_file(self, path: Path):
-        # TODO: Using get_include_files here isn't ideal; if the user puts in a glob that
-        # matches files but doesn't match any files in context, we won't know what that glob is
-        # and can't return it as an invalid path
-        paths, invalid_paths = get_include_files([path], [])
-        removed_paths = list[Path]()
-        for new_path in paths.keys():
-            if new_path in self.include_files:
-                removed_paths.append(new_path)
-                del self.include_files[new_path]
-        return removed_paths, invalid_paths
+    def include(
+        self, path: Path | str, ignore_patterns: Iterable[Path | str] = []
+    ) -> Set[Path]:
+        """Add code to the context
+
+        '.' is replaced with '*' (recusively search the cwd)
+
+        Args:
+            `path`: can be a relative or absolute file path, file interval path, directory, or glob pattern.
+
+        Return:
+            A set of paths that have been successfully added to the context
+        """
+        session_context = SESSION_CONTEXT.get()
+
+        path = Path(path)
+        if str(path) == ".":
+            path = Path("*")
+
+        included_paths: Set[Path] = set()
+        try:
+            code_features = get_code_features_for_path(
+                path=path,
+                cwd=session_context.cwd,
+                ignore_patterns=[
+                    *ignore_patterns,
+                    *self.ignore_patterns,
+                    *session_context.config.file_exclude_glob_list,
+                ],
+            )
+        except PathValidationError as e:
+            session_context.stream.send(e, color="light_red")
+            return included_paths
+
+        for code_feature in code_features:
+            if code_feature.path in self.include_files:
+                self.include_files[code_feature.path].append(code_feature)
+            else:
+                self.include_files[code_feature.path] = [code_feature]
+
+        return included_paths
+
+    def exclude(self, path: Path | str) -> Set[Path]:
+        """Remove code from the context
+
+        # NOTE: should this be the intended behavior?
+        Code Features are removed from the context and are excluded from any future `include` calls (unless explicity specified).
+        For paths that are files or file intervals, if they don't already exist in the Code Context an Exception is thrown.
+        For file intervals, if the interval doesn't *exactly match* any intervals in the code context, nothing is removed.
```

Comment:
Would like feedback/thoughts on this

Reply:
I don't like this behavior; I think exclude should just remove the file (or all files in the directory, or all matching files for a glob pattern, etc.) whenever it gets called. Otherwise, we're going to end up with some pretty unintuitive behavior (you /exclude'd a directory? Suddenly nothing is ever being added in when you try to include it later).

Reply:
Yeah I agree w/ that. I actually never added the logic that excludes any file/patterns from future `include` calls, so I just won't implement that